nixos/lib/make-zfs-image.nix

```diff
@@ -0,0 +1,340 @@
+  bootSize ? 1024
+, # features on the boot pool
+  # Note: the pool has no features enabled by default,
+  # and only this list of features is enabled. This list
+  # must match GRUB's supported list.
+  # See: https://github.com/openzfs/openzfs-docs/blame/master/docs/Getting%20Started/Debian/Debian%20Stretch%20Root%20on%20ZFS.rst#L200-L216
+  bootPoolFeatures ? [
+    "async_destroy"
+    "bookmarks"
+    "embedded_data"
+    "empty_bpobj"
+    "enabled_txg"
+    "extensible_dataset"
+    "filesystem_limits"
+    "hole_birth"
+    "large_blocks"
+    "lz4_compress"
+    "spacemap_histogram"
+    "userobj_accounting"
+  ]
+
+, # zpool properties
+  bootPoolProperties ? {
+    autoexpand = "on";
+  }
+, # pool-wide filesystem properties
+  bootPoolFilesystemProperties ? {
+    acltype = "posixacl";
+    atime = "off";
+    compression = "on";
+    mountpoint = "legacy";
+    xattr = "sa";
+  }
+, # The name of the ZFS pool
+  rootPoolName ? "tank"
+
+, # zpool properties
+  rootPoolProperties ? {
+    autoexpand = "on";
+  }
+, # pool-wide filesystem properties
+  rootPoolFilesystemProperties ? {
+    acltype = "posixacl";
+    atime = "off";
+    compression = "on";
+    mountpoint = "legacy";
+    xattr = "sa";
+  }
+
+, # datasets, with per-attribute options:
+  # mount: (optional) mount point in the VM
+  # properties: (optional) ZFS properties on the dataset, like filesystemProperties
+  # Note: datasets will be created from shorter to longer names as a simple topo-sort
+  datasets ? {
+    "${rootPoolName}/system/root".mount = "/";
+    "${rootPoolName}/system/var".mount = "/var";
+    "${rootPoolName}/local/nix".mount = "/nix";
+    "${rootPoolName}/user/home".mount = "/home";
+    "${bootPoolName}/boot".mount = "/boot";
+  }
+
+, # The files and directories to be placed in the target file system.
+  # This is a list of attribute sets {source, target} where `source'
+  # is the file system object (regular file or directory) to be
+  # grafted in the file system at path `target'.
+  contents ? []
+
+, # The initial NixOS configuration file to be copied to
+  # /etc/nixos/configuration.nix. This configuration will be embedded
+  # inside a configuration which includes the described ZFS fileSystems.
+  configFile ? null
+
+, # Shell code executed after the VM has finished.
+  postVM ? ""
+
+, name ? "nixos-disk-image"
+
+, # Disk image format, one of qcow2, qcow2-compressed, vdi, vpc, raw.
+  format ? "raw"
+
+, # Include a copy of Nixpkgs in the disk image
+  includeChannel ? true
+}:
+let
+  formatOpt = if format == "qcow2-compressed" then "qcow2" else format;
+
+  compress = lib.optionalString (format == "qcow2-compressed") "-c";
+
+  filenameSuffix = "." + {
+    qcow2 = "qcow2";
+    vdi = "vdi";
+    vpc = "vhd";
+    raw = "img";
+  }.${formatOpt} or formatOpt;
+  bootFilename = "nixos.boot${filenameSuffix}";
+  rootFilename = "nixos.root${filenameSuffix}";
+
+  # FIXME: merge with channel.nix / make-channel.nix.
```

Comment:
FIXME when?

Reply:
Your guess is as good as mine? :) https://github.com/NixOS/nixpkgs/commit/9802da517fe4da41d68516029b01c5c1f175b3e4 :grimacing: